### PR TITLE
Server & game status detection

### PR DESCRIPTION
**Pull request by dadodasyra**
It may still need modification in the main.rs to support javascript bridging with tauri.
I tried but it's a mess and don't have time for it.

The whole API behind is working, tried by running rust without tauri.

It needs Administrator priveleges

### CODE REVIEW

File: frontend/src-tauri/src/api.rs [hunk 1]
```diff
@@ -0,0 +1,59 @@
+use std::time::Instant;
+use serde::Serialize;
+
+// Assuming these are your global variables in the `api` module
```
Comment (zelytra) [inline frontend/src-tauri/src/api.rs:4]: Message GPT

File: frontend/src-tauri/src/api.rs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+#[derive(Clone)]
+pub struct Api {
+    pub game_status: GameStatus,
+    pub server_ip: String,
+    pub server_port: u16,
+    pub last_updated_server_ip: Instant
+}
+
+#[derive(PartialEq, Debug, Clone, Serialize)]
+pub enum GameStatus {
```
Comment (zelytra) [inline frontend/src-tauri/src/api.rs:14]: Ajouter des descriptif de chaque etat

File: frontend/src-tauri/src/api.rs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+    /**
+    * Server IP, may be updated
+    */
+    pub async fn get_server_ip(&self) -> String {
+        self.server_ip.clone()
+    }
+
+
+    pub async fn get_server_port(&self) -> u16 {
```
Comment (zelytra) [inline frontend/src-tauri/src/api.rs:50]: Ajouter doc

File: frontend/src-tauri/src/api.rs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        self.server_port
+    }
+
+
+    pub async fn get_last_updated_server_ip(&self) -> Instant {
```
Comment (zelytra) [inline frontend/src-tauri/src/api.rs:55]: Ajouter doc

File: frontend/src-tauri/src/fetch_informations.rs [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+
+
+pub async fn init() -> std::result::Result<Arc<RwLock<Api>>, anyhow::Error> {
+    let api_base = Arc::new(RwLock::new(Api::new()));
+    let api  = Arc::clone(&api_base);
+
+    tokio::spawn(async move {
+        loop {
+            let pid = find_pid_of("SoTGame.exe");
+            //println!("PID: {:?}", pid);
```
Comment (zelytra) [inline frontend/src-tauri/src/fetch_informations.rs:32]: Debug à retirer

File: frontend/src-tauri/src/fetch_informations.rs [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+
+            if pid.is_empty() {
+                api.write().await.game_status = GameStatus::Closed;
+            } else {
+                let pid = pid[0].parse().unwrap();
+                let udp_connections = get_udp_connections(pid);
+                //println!("UDP connections: {:?}", udp_connections);
```
Comment (zelytra) [inline frontend/src-tauri/src/fetch_informations.rs:39]: Debug à retirer

File: frontend/src-tauri/src/fetch_informations.rs [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+
+                    let ip_addresses = match hostname.to_socket_addrs() { //TODO Optimization: Cache IP ADDRESSES
+                        Ok(addrs) => addrs.map(|socket_addr| socket_addr.ip()).collect::<Vec<IpAddr>>(),
+                        Err(e) => {
+                            eprintln!("Error getting IP addresses: {}", e);
+                            continue;
+                        }
+                    };
+
+                    println!("IP addresses: {:?}", ip_addresses);
```
Comment (zelytra) [inline frontend/src-tauri/src/fetch_informations.rs:73]: Debug à retirer

File: frontend/src-tauri/src/fetch_informations.rs [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+
+                                    // Release the lock
+                                    drop(api_lock);
+                                }
+
+                                None => {
+                                    let last_updated_server_ip = api_clone.read().await.last_updated_server_ip;
+
+                                    // When no result since > 20 seconds reset server_ip
+                                    // This should never happen since when the player is not connected to a server, the game_status is MainMenu
```
Comment (zelytra) [inline frontend/src-tauri/src/fetch_informations.rs:106]: Mauvais endroit du commentaire

File: frontend/src-tauri/src/fetch_informations.rs [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+                                }
+                            }
+                        });
+                    }
+                }
+            }
+
+            let game_status = api.read().await.game_status.clone();
+            let dynamic_time = match game_status {
+                GameStatus::Closed => 5000,
```
Comment (zelytra) [inline frontend/src-tauri/src/fetch_informations.rs:129]: Voir si possibilité de descendre sous la secondes voir les 500ms

File: frontend/src-tauri/src/fetch_informations.rs [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+    for line in reader.lines().filter_map(|l| l.ok()) {
+        if let Ok(port) = line.parse::<u16>() {
+            ports.push(port);
+        }
+    }
+
+    ports
+}
+
+fn get_local_hostname() -> std::io::Result<String> {
```
Comment (zelytra) [inline frontend/src-tauri/src/fetch_informations.rs:170]: A documenter

File: frontend/src-tauri/src/fetch_informations.rs [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+    Ok(hostname::get()?.into_string().unwrap_or_else(|_| "localhost".into()))
+}
+
+pub fn find_pid_of(process_name: &str) -> Vec<String> {
```
Comment (zelytra) [inline frontend/src-tauri/src/fetch_informations.rs:174]: A documenter

File: frontend/src-tauri/src/fetch_informations.rs [hunk 12]
```diff
@@ -0,0 +1,342 @@
...
+            // Timeout reached without receiving a packet
+            return None;
+        }
+
+        tokio::select! {
+            recv_result = socket.recv(&mut buf) => {
+                match recv_result {
+                    Ok(len) if len > 0 => {
+                        let recv_result = socket.recv(&mut buf).await;
+                        //println!("{:?}", recv_result);
```
Comment (zelytra) [inline frontend/src-tauri/src/fetch_informations.rs:204]: Debug a retirer

File: frontend/src-tauri/src/fetch_informations.rs [hunk 13]
```diff
@@ -0,0 +1,342 @@
...
+    let timeout = Duration::from_millis(3000);
+    let start_time = Instant::now();
+
+    loop {
+        if start_time.elapsed() > timeout {
+            // Timeout reached without receiving a packet
+            return None;
+        }
+
+        tokio::select! {
```
Comment (zelytra) [inline frontend/src-tauri/src/fetch_informations.rs:199]: A documenter

**Comment by dadodasyra** [comment]
```ts
import { invoke } from '@tauri-apps/api/tauri'

setInterval(() => {
    invoke('get_game_status')
        .then((res) => {
                console.log(res);
                if (res === "InGame") {
                    invoke('get_server_ip')
                        .then((res) =>
                            console.log(res)
                        )
                }
        }).catch((e) => console.error(e))
}, 5000)

```
Get data example

File: frontend/src-tauri/src/fetch_informations.rs [hunk 14]
```diff
@@ -0,0 +1,342 @@
...
+            recv_result = socket.recv(&mut buf) => {
+                match recv_result {
+                    Ok(len) if len > 0 => {
+                        let recv_result = socket.recv(&mut buf).await;
+                        //println!("{:?}", recv_result);
+                        return match recv_result {
+                            Ok(len) => {
+                                let packet = PacketHeaders::from_ip_slice(&buf[0..len]).ok()?;
+                                //println!("{:?}", packet.transport);
```
Comment (zelytra) [inline frontend/src-tauri/src/fetch_informations.rs:208]: Debug a retirer

**Review by zelytra — CHANGES_REQUESTED** [review]

**Review by zelytra — APPROVED** [review]

**Pull request merged**